tsfc/driver.py

```diff
@@ -173,75 +172,73 @@ def compile_integral(idata, fd, prefix, parameters):
                                                    quad_rule.points)
 
         integrand = fem.replace_coordinates(integral.integrand(), coordinates)
-        quadrature_index = ein.Index(name="ip%d" % i)
+        quadrature_index = gem.Index(name="ip%d" % i)
         quadrature_indices.append(quadrature_index)
         nonfem = fem.process(integral_type, integrand,
                              tabulation_manager, quad_rule.weights,
                              quadrature_index, argument_indices,
                              coefficient_map, index_cache)
         nonfem = opt.unroll_indexsum(nonfem, max_extent=3)
-        nonfem_.append([(ein.IndexSum(e, quadrature_index) if quadrature_index in e.free_indices else e)
+        nonfem_.append([(gem.IndexSum(e, quadrature_index) if quadrature_index in e.free_indices else e)
                         for e in nonfem])
 
     # Sum the expressions that are part of the same restriction
-    nonfem = list(reduce(ein.Sum, e, ein.Zero()) for e in zip(*nonfem_))
+    nonfem = list(reduce(gem.Sum, e, gem.Zero()) for e in zip(*nonfem_))
     simplified = opt.remove_componenttensors(nonfem)
 
-    cell_orientations = False
-    refcount = sch.count_references(simplified)
-    candidates = set()
+    # Look for cell orientations in the simplified GEM
+    kernel.oriented = False
     for node in traversal(simplified):
-        if isinstance(node, ein.IndexSum):
-            if refcount[node.children[0]] == 1:
-                candidates.add(node.children[0])
-        else:
-            for child in node.children:
-                if set(child.free_indices) == set(node.free_indices) and refcount[child] == 1:
-                    if not (isinstance(child, ein.Literal) and child.shape):
-                        candidates.add(child)
-
-        if isinstance(node, ein.Variable) and node.name == "cell_orientations":
-            cell_orientations = True
-
-    if cell_orientations:
-        decl = coffee.Decl("int *restrict *restrict", coffee.Symbol("cell_orientations"),
-                           qualifiers=["const"])
-        arglist.insert(2, decl)
-        kernel.oriented = True
-
-    # Need a deterministic ordering for these
-    indices = set()
+        if isinstance(node, gem.Variable) and node.name == "cell_orientations":
+            kernel.oriented = True
+            decl = coffee.Decl("int *restrict *restrict",
+                               coffee.Symbol("cell_orientations"),
+                               qualifiers=["const"])
+            arglist.insert(2, decl)
+            break
+
+    # Collect indices in a deterministic order
+    indices = []
     for node in traversal(simplified):
-        indices.update(node.free_indices)
-    indices = sorted(indices)
+        if isinstance(node, gem.Indexed):
+            indices.extend(node.multiindex)
+    _, unique_indices = numpy.unique(indices, return_index=True)
+    indices = numpy.asarray(indices)[sorted(unique_indices)]
```

Comment:
OK, so you're computing the unique indices, preserving the order.  Use np.sort(a) to avoid casting back and forth to list?